### ISSUE TITLE
CPU INT8 quantization

### ISSUE THREAD

**Issue by feifeibear** [Tencent/TurboTransformers#103]
Using FBGEMM to support CPU quantization.

**Comment by feifeibear** [comment]
Motivation: we use FBGEMM in order to have consistent accuracy as PyTorch dynamic quantization.
As TurboTransformer's optimizations are focused on Non-GEMM operations, we can reuse PyTorch QLinear code as much as possible.
The most ideal way is to call Pytorch dynamic quantization APIs at Python level.
One significant shortcoming of this method is that it will constantly convert tensor from dlpack tensor format to Pytorch tensor format.
If it doesn't work, we have to reproduce Pytorch QLinear Layer in Turbo.
@flans39

**Comment by feifeibear** [comment]
I wrote a test script to understand Pytorch dynamic quantization.
```
import torch
from torch import nn

torch.set_grad_enabled(False)
LinearLayer = torch.nn.Linear(20, 30)
model = nn.Sequential(LinearLayer)


# for reference
qLinearLayer2 = torch.nn.quantized.dynamic.Linear(20, 30)
print(type(qLinearLayer2))
print(qLinearLayer2._packed_params._packed_params)


qmodel = torch.quantization.quantize_dynamic(model, {torch.nn.Linear}, dtype=torch.qint8)
# qLinearLayer = torch.quantization.quantize_dynamic(LinearLayer)
print(qmodel)
print(type(qmodel))
# print(qLinearLayer._packed_params._packed_params)

input = torch.rand(5, 20)
# print("input =", input)

res = model(input)
print("res =", res)

qres = qmodel(input)
print("qres =", qres)
```

**Comment by flans39** [comment]
> Motivation: we use FBGEMM in order to have consistent accuracy as PyTorch dynamic quantization.
> As TurboTransformer's optimizations are focused on Non-GEMM operations, we can reuse PyTorch QLinear code as much as possible.
> The most ideal way is to call Pytorch dynamic quantization APIs at Python level.
> One significant shortcoming of this method is that it will constantly convert tensor from dlpack tensor format to Pytorch tensor format.
> If it doesn't work, we have to reproduce Pytorch QLinear Layer in Turbo.
> @flans39

I've reviewed PyTorch's implementation of dynamic quantization. There are three levels of quantization APIs in PyTorch: `torch.quantization` includes high-level tools of quatization; `torch.nn.quantization` provides the quantized version of nn.Module; low-level APIs includes quantized tensors and operations.

For example, in the above test script, `quantize_dynamic()` transfers the `nn.Linear` layers of the model into `nn.quantized.dynamic.Linear` layers. Then, at the time of inference, the follwing functions are invoked:

1. `torch.nn.quantized.dynamic.Linear.forward()`
2. `torch.ops.quantized.linear_dynamic`
3. C++ interface
4. `QLinearDynamicInt8::run`
5. `PackedLinearWeight::apply_dynamic`
6. `PackedLinearWeight::apply_dynamic_impl`

where the C++ interface is defined in `pytorch/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp`
```
TORCH_LIBRARY_IMPL(quantized, CPU, m) {
  m.impl("linear_dynamic", TORCH_FN(QLinearDynamicInt8<false>::run));
  m.impl("linear_relu_dynamic", TORCH_FN(QLinearDynamicInt8<true>::run));
  m.impl("linear_dynamic_fp16", TORCH_FN(QLinearDynamicFp16<false>::run));
}
```

On x86 systems, PyTorch takes FBGEMM as the backend of `PackedLinearWeight::apply_dynamic_impl`. The entry point is
```
        fbgemm::fbgemmPacked(
            /*packA=*/packA,
            /*packB=*/*packB,
            /*C=*/output.data_ptr<float>(),
            /*C_buffer=*/buffer.data_ptr<int32_t>(),
            /*ldc=*/N,
            /*outProcess=*/outputProcObj,
            /*thread_id=*/task_id,
            /*num_threads=*/num_tasks);
```

To support quantization in TurboTransformers, we can reuse high-level tools in `torch.quantization`, integrate the quantized modules in `torch.nn.quantization` into `python/turbo_transformers/layers` and reproduce low-level quantized tensors and operations.

**Comment by feifeibear** [comment]
> > Motivation: we use FBGEMM in order to have consistent accuracy as PyTorch dynamic quantization.
> > As TurboTransformer's optimizations are focused on Non-GEMM operations, we can reuse PyTorch QLinear code as much as possible.
> > The most ideal way is to call Pytorch dynamic quantization APIs at Python level.
> > One significant shortcoming of this method is that it will constantly convert tensor from dlpack tensor format to Pytorch tensor format.
> > If it doesn't work, we have to reproduce Pytorch QLinear Layer in Turbo.
> > @flans39
> 
> I've reviewed PyTorch's implementation of dynamic quantization. There are three levels of quantization APIs in PyTorch: `torch.quantization` includes high-level tools of quatization; `torch.nn.quantization` provides the quantized version of nn.Module; low-level APIs includes quantized tensors and operations.
> 
> For example, in the above test script, `quantize_dynamic()` transfers the `nn.Linear` layers of the model into `nn.quantized.dynamic.Linear` layers. Then, at the time of inference, the follwing functions are invoked:
> 
> 1. `torch.nn.quantized.dynamic.Linear.forward()`
> 2. `torch.ops.quantized.linear_dynamic`
> 3. C++ interface
> 4. `QLinearDynamicInt8::run`
> 5. `PackedLinearWeight::apply_dynamic`
> 6. `PackedLinearWeight::apply_dynamic_impl`
> 
> where the C++ interface is defined in `pytorch/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp`
> 
> ```
> TORCH_LIBRARY_IMPL(quantized, CPU, m) {
>   m.impl("linear_dynamic", TORCH_FN(QLinearDynamicInt8<false>::run));
>   m.impl("linear_relu_dynamic", TORCH_FN(QLinearDynamicInt8<true>::run));
>   m.impl("linear_dynamic_fp16", TORCH_FN(QLinearDynamicFp16<false>::run));
> }
> ```
> 
> On x86 systems, PyTorch takes FBGEMM as the backend of `PackedLinearWeight::apply_dynamic_impl`. The entry point is
> 
> ```
>         fbgemm::fbgemmPacked(
>             /*packA=*/packA,
>             /*packB=*/*packB,
>             /*C=*/output.data_ptr<float>(),
>             /*C_buffer=*/buffer.data_ptr<int32_t>(),
>             /*ldc=*/N,
>             /*outProcess=*/outputProcObj,
>             /*thread_id=*/task_id,
>             /*num_threads=*/num_tasks);
> ```
> 
> To support quantization in TurboTransformers, we can reuse high-level tools in `torch.quantization`, integrate the quantized modules in `torch.nn.quantization` into `python/turbo_transformers/layers` and reproduce low-level quantized tensors and operations.

Can you paste some preliminary benchmarking results on PyTorch dynamic quantization?

**Comment by flans39** [comment]
> Can you paste some preliminary benchmarking results on PyTorch dynamic quantization?

Here it is. (on a virtual 8-core CPU @2.6 GHz)

M, N, K | Threads | torch fp32 GOPS | torch int8 GOPS
-- | -- | -- | --
128, 128, 128 | 1 | 21.3 | 61.3
128, 128, 128 | 2 | 36.9 | 96.8
128, 128, 128 | 4 | 59.1 | 134.5
128, 128, 128 | 8 | 51.9 | 102.2
256, 512, 256 | 1 | 24.6 | 90.3
256, 512, 256 | 2 | 47.9 | 169.3
256, 512, 256 | 4 | 88.1 | 292.7
256, 512, 256 | 8 | 89.4 | 270.2
1024, 1024, 1024 | 1 | 25.0 | 93.7
1024, 1024, 1024 | 2 | 48.6 | 182.8
1024, 1024, 1024 | 4 | 95.0 | 340.5
1024, 1024, 1024 | 8 | 100.0 | 340.9
